src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -201,6 +203,7 @@ protected internal static void ValidateMatchTimeout(TimeSpan matchTimeout)
         /// Regex constructor, we don't load RegexCompiler and its reflection classes when
         /// instantiating a non-compiled regex.
         /// </summary>
+        [RequiresDynamicCode("The native code for Regex compilation might not be available at runtime.")]
```

Comment:
> In a perfect world, the UnconditionalSuppressMessage attribute on the internal Regex constructor shouldn't even be needed because it only calls this method behind a check for RuntimeFeature.IsDynamicCodeCompiled. The analyzer should be able to tell that the method that "RequiresDynamicCode" doesn't ever get invoked when dynamic code isn't available.

Right, we talked about that on the other PRs/issues.  I was asking here about, short of recognizing IsDynamicCodeCompiled checks in that manner, it could do the same for UnconditionalSuppressMessage.  Both would be valuable, and presumably both require the same level of cross-method analysis (recursively suppressing diagnostics on methods only callable from call sites suppressed via either a guard check or attribute).

Reply:
> and presumably both require the same level of cross-method analysis 

The check I'm taking about shouldn't require the same level of cross-method analysis. We would still annotate the `Compile` method as `RequiresDynamicCode`. The analyzer wouldn't need any more cross-method analysis than it does today - it just looks at the method being called and looks at its attributes. This would be the same behavior as the SupportedOSPlatform analyzer we have today - it doesn't warn about code inside `if (OperatingSystem.IsWindows())` that calls windows-only code.

Reply:
> We would still annotate the Compile method as RequiresDynamicCode

And we don't want to auto-propagate annotations up the call chain because... it's too expensive for the analyzer?

Reply:
The authors of the analysis tools would need to answer that (@vitek-karas @MichalStrehovsky @agocke). But note that this is also the way the SupportedOSPlatform analyzer works - you need to annotate the entire call chain. So it isn't limited to the RequiresDynamicCode and RequiresUnreferencedCode analyzers.

Reply:
I presume every public entry point would be considered a use in this model?

I think it would probably be computationally infeasible for the analyzer, but I also don't think it's a good UX. The impact is too far separate from the effect to be predictable by the user. You would get weird effects, like making a private method public, and suddenly it starts warning, because when it calls "A->B->C->D", D is annotated.

In general, I think global graph analysis sounds nice, but it's bad to debug.

Reply:
>  I also don't think it's a good UX

That may be. I think the same thing about having to add a bunch of attributes to suppress warnings that, from my perspective, are bogus.

Reply:
Also, the point of the `requires` attributes is also to tell developers -- these methods have special requirements. It's an API demand, so it should be visible as part of the API of the method.

Reply:
> It's an API demand, so it should be visible as part of the API of the method.

For exposed API, sure. None of the ones in this PR are.

Reply:
Propagating these across methods is problematic because e.g. these annotations need to be consistent across virtuals (if base is not annotated, override cannot be annotated)

It's hard to imagine a good UX around situations when the propagation stops working at some point in the static callgraph. There can be big ripple effects. Also scenarios like adding InternalsVisibleTo to the assembly popping up warnings, etc.